### PR TITLE
feat: improved logging

### PR DESCRIPTION
**Pull request by blu3beri**
- Custom own logger macros
- Able to log from every crate (this should be reworked when we want to add a TUI so we can log with a notification for example)
- No rework of the logs themselves yet (will come in the future)

### CODE REVIEW

File: crates/logger/src/macros.rs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+#[macro_export]
+macro_rules! log_debug {
+    ($($arg:tt)+) => {
+        if crate::logger::STATE.read().unwrap().level == crate::logger::LogLevel::Debug {
+            println!("[DEBUG]: {}", format!($($arg)+));
+        }
+    };
+}
+
+/// Simple trace logger
```
Comment (morrieinmaas) [inline crates/logger/src/macros.rs:38]: This seems quite generic/repetitive how we do the logging macros here. Maybe this can be refactored?

**Review by morrieinmaas — APPROVED** [review]
LGTM! overall. maybe the macros can be refactored? 

Nice job

**Comment by blu3beri** [comment]
Screenshot 2022-05-29 at 14 25 27
Screenshot 2022-05-29 at 14 25 05
Screenshot 2022-05-29 at 14 24 41
Screenshot 2022-05-29 at 14 24 12


couple of examples of the logger.

@morrieinmaas I fixed your feedback.

**Comment by blu3beri** [comment]
@jloleysens I removed the success logger as it was very redundant and info should suffice.

File: crates/logger/src/macros.rs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+/// Simple wrapper around println!
+#[macro_export]
+macro_rules! log {
+    ($($arg:tt)*) => {
+        if crate::logger::STATE.read().unwrap().level != crate::logger::LogLevel::Off {
+            println!($($arg)*);
+        }
+    };
+}
```
Comment (jloleysens) [inline crates/logger/src/macros.rs:35]: I like this distinction between "just print something" and "print something at a log level".

This gives use the ability to separate logs that form part of the UX that we always want to show and logs that are about things like errors, warnings or general info.

File: crates/cli/src/modules/connection.rs [hunk 3]
```diff
@@ -122,14 +120,14 @@ pub async fn parse_connection_args(
             };
             agent.create_invitation(options).await.map(|response| {
                 loader.stop();
-                info!("{} invite with connection id: ", "Created".green());
-                println!("{}", response.connection_id);
+                log_info!("{} invite with connection id: ", "Created".green());
```
Comment (jloleysens) [inline crates/cli/src/modules/connection.rs:123]: We spoke about this briefly, and I think it is totally OK to use colors inside levelled logs 👍🏻

File: crates/cloudagent-python/src/web.rs [hunk 4]
```diff
@@ -55,29 +57,30 @@ impl CloudAgent {
...
 
-        trace!("About to send request:\n{:#?}", client);
+        log_trace!("About to send request:");
+        log_trace!("{:#?}", client);
         match client.send().await {
             Ok(res) => {
                 let status_code = res.status().as_u16();
-                trace!("Got {} response:\n{:#?}", status_code, res);
+                log_info!("Got {} response:", status_code);
+                log_info!("{:#?}", res);
```
Comment (jloleysens) [inline crates/cloudagent-python/src/web.rs:66]: I think this is pretty verbose for "info" level output.

```suggestion
                log_trace!("Got {} response:", status_code);
                log_trace!("{:#?}", res);
```

File: crates/cli/Cargo.toml [hunk 5]
```diff
@@ -16,5 +17,3 @@ serde = { version = "1.0.130", features = ["derive"] }
 serde_json = "1.0.68"
 serde_yaml = "0.8.23"
 tokio = { version = "1", features = ["full"] }
-simplelog = "0.11.2"
-log = "0.4.14"
```
Comment (jloleysens) [inline crates/cli/Cargo.toml:20]: One other approach could have been to keep the Rust `log` facade (basically a set of macros) and implement a logger that the facade uses, per https://docs.rs/log/latest/log/#implementing-a-logger.

Comment (jloleysens) [reply crates/cli/Cargo.toml:20]: In that case we would not have our own macros except perhaps `log!` for `println!` and could leave current uses of `info!`, `trace!` etc.

**Review by jloleysens — COMMENTED** [review]
Great work @blu3beri , left a couple of observations based on what I read in the code.

I think we should keep the current network responses as a `trace` level log however, WDYT?

Screenshot 2022-06-01 at 11 34 44

FWIW, I also think the green "Created" looks 🆗 with the new log format.

File: crates/cli/src/modules/connection.rs (hunk 3 above)
Comment (blu3beri) [reply crates/cli/src/modules/connection.rs:123]: Yeah it probably okay, just worried about consistency and which colors for what verbs etc.

File: crates/cli/Cargo.toml (hunk 5 above)
Comment (blu3beri) [reply crates/cli/Cargo.toml:20]: I am not too sure what that would add. I think the code we have now is small enough and adds a lot of customizability, e.g. making a log without a newline, etc.

Comment (blu3beri) [reply crates/cli/Cargo.toml:20]: I will merge this for now and if we want to revisit this in the future I am okay with that.

**Pull request merged**